### PR TITLE
Improvements to documentation in arm9/math.h

### PR DESCRIPTION
**Pull request by awiebe**
I've started work on clearing up some of the finer points of the shall we say... terse API documentation.

I'm adding preambles with relevant information, correcting errors and adding clarifications.

### CODE REVIEW

File: include/nds/arm9/math.h [hunk 1]
```diff
@@ -68,9 +83,10 @@
 
 //  Fixed point conversion macros
 
-#define inttof32(n)          ((n) << 12) /*!< \brief convert int to f32 */
-#define f32toint(n)          ((n) >> 12) /*!< \brief convert f32 to int */
-#define floattof32(n)        ((int)((n) * (1 << 12))) /*!< \brief convert float to f32 */
+#define inttof32(n)          ((n) << 12) /*!< \brief convert a 32 bit integer to a 20.12 fixed point(f32) */
+#define f32toint(n)          ((n) >> 12) /*!< \brief convert a 20.12 fixed point(f32) to a 32 bit integer, the fractional portion is truncated. */
+#define floattof32(n)        ((int)((n) * (1 << 12))) /*!< \brief convert float to f32 (20.12)
+\warn The sign bit will be lost becuase it is the MSB which is truncated  */
```
Comment (AntonioND) [inline include/nds/arm9/math.h]: Really? I don't remember it being like that, and I don't understand how the sign may be lost after casting to int and multiplying.

**Comment by WinterMute** [comment]
Sorry, this is mostly wrong. Fixed point doesn't work like this at all.

These fixed point numbers are 20.12, end of story. There is no 1.19.12 format and the int cast does not truncate the "sign bit".

Fixed point is simply a way of using integer arithmetic to represent fractions by multiplying & dividing by a constant power of 2 factor. I'll see if I can dig out some notes and articles for you that explain this stuff.

It's both simpler and more complex than you think.

**Comment by awiebe** [comment]
`crossf32()` , `dotf32() `and `normalizef32()` all make reference to the signed format of 1.19.12 in their description, not in my revisions, but in the origin/master. So either there is such a format... or that documentation is just flat out wrong.

I have reason to believe 1.19.12 is a signed magnitude fixed point number, where some functions are just unsigned and use the 20.12 representation, the actual in memory representation doesn't change, like a signed and unsigned integer, but the way the data paths handle it changes its interpretation.

If you start with a signed integer and shift it i<<12 to convert that to a fixed point the MSB will be lost... come to think of it that's considerable overflow and you'll have a bigger problem then just losing sign, I'll change it to say overflow, integer data in the upper 12 bits will be lost.

The macro for float conversion completely ignores the sign bit and fractional portion , I know there are only soft floating points so maybe it's doing some sorcery I don't understand, but it seems to me it's just casting it down to integer and then shifting it.

As I understand it a fixed point number is represented as follows:
1.m.n
- 1 the MSB is the sign bit, if omitted, it's unsigned
- m the next m bit represent the whole number portion as if it were just an integer
- n the fractional portion forms an integer portion where them MSB has the greatest negative magnitude

So a 1.11.4 fixed point would be
s|wwwwwwwwww|ffff
```
-1^(s-1) ( w*2^12+ ...  +w*(2^4)+ w*(2^3)+ w*(2^2)+ w*(2^1)+ w*(2^0) 
+ f*(2^-1) + f*(2^-2) + f*(2^-3) + f*(2^-4) )
```

**Comment by WinterMute** [comment]
That documentation is just flat out wrong ...

**Comment by AntonioND** [comment]
@WinterMute is right, you don't understand how it works. It's true that sometimes doing a left shift to a value to multiply by a power of 2 will make you lose the sign, but that's because it is an overflow! A regular multiplication has the same problem, and you wouldn't say it uses sign-magnitude format. It's the normal behaviour.

The fixed point format used in here is just a simple 32-bit two's-complement signed integer in which we consider the lower 12 bits to be the fractionary part, there's nothing more to it.

So yes, the documentation in the master branch is wrong, but that doesn't make yours right.

**Comment by awiebe** [comment]
Ok so then that particular documentation is wrong and the article on nds development i was reading is wrong.

So just to be sure all nds fixed point numbers are described as
Some 2's complement integer m.n where **n is the number of places from the left where the decimal point resides arbitrarily base-2**

For 10.12 this would be
(ddddddddddd.ffffffffffff)_2

I'll start by changing the documentation on the fixed point vector functions so they represent 2'sc instead of signed magnitude.  Then try modifying the other docs individually until they reflect this new information.

**Comment by WinterMute** [comment]
https://www.coranac.com/tonc/text/fixed.htm is a reasonable grounding in fixed point stuff.

It says that technically speaking, the 1.n.f documentation is false which I'd disagree with - it's just false. There are processors which **do** have a sign bit and of course actual floats use a signed magnitude format. Implying that fixed point has this format is just plain wrong.

**Pull request closed without merging**